File: 1448-CountGoodNodesinBinaryTree/1448-CountGoodNodesinBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {

    function dfs(node, acc){
        if(node == null) return 0;

        acc = Math.max(acc, node.val)

        let ans = dfs(node.left, acc) + dfs(node.right, acc)
        
        if(node.val >= acc) {
            ans++;
        }
        return ans;
    }
    return dfs(root, -Infinity);
};

File: 1448-CountGoodNodesinBinaryTree/1448-CountGoodNodesinBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(root, acc):
            if root == None: return 0

            acc = max(acc, root.val)

            if acc <= root.val:
                return 1 + dfs(root.left, acc) + dfs(root.right, acc)
            else:
                return dfs(root.left, acc) + dfs(root.right, acc)




        return dfs(root, float('-inf'))


        